### PR TITLE
[SPARK-48307][SQL] InlineCTE should keep not-inlined relations in the original WithCTE node

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

I noticed an outdated comment in the rule `InlineCTE`
```
      // CTEs in SQL Commands have been inlined by `CTESubstitution` already, so it is safe to add
      // WithCTE as top node here.
```

This is not true anymore after https://github.com/apache/spark/pull/42036 . It's not a big deal as we replace not-inlined CTE relations with `Repartition` during optimization, so it doesn't matter where we put the `WithCTE` node with not-inlined CTE relations, as it will disappear eventually. But it's still better to keep it at its original place, as third-party rules may be sensitive about the plan shape.

### Why are the changes needed?

to keep the plan shape as much as can after inlining CTE relations.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @viirya @amaliujia @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala [hunk 1]
```diff
@@ -74,34 +70,33 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
...
         buildCTEMap(child, cteMap, outerCTEId)
 
       case ref: CTERelationRef =>
-        val (cteDef, refCount, refMap) = cteMap(ref.cteId)
-        cteMap(ref.cteId) = (cteDef, refCount + 1, refMap)
+        cteMap(ref.cteId) = cteMap(ref.cteId).withRefCountIncreased(1)
         outerCTEId.foreach { cteId =>
-          val (_, _, outerRefMap) = cteMap(cteId)
-          outerRefMap(ref.cteId) += 1
+          cteMap(cteId).recordOutgoingReference(ref.cteId)
```
Comment (amaliujia) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala:99]: I personally like to not hide the `+1` implementation details here as I had troubles to understand it.

Or the function name can be `IncreaseOutgoingReferenceByOne`?

**Review by amaliujia — APPROVED** [review]
LGTM

**Review by LuciferYang — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTESuite.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.{AppendData, CTERelationDef, CTERelationRef, LogicalPlan, OneRowRelation, WithCTE}
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
+
+class InlineCTESuite extends PlanTest {
+
+  object Optimize extends RuleExecutor[LogicalPlan] {
+    val batches = Batch("inline CTE", FixedPoint(100), InlineCTE()) :: Nil
+  }
+
+  test("not-inlined CTE relation in command") {
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTESuite.scala:33]:
```suggestion
  test("SPARK-48307: not-inlined CTE relation in command") {
```

**Comment by viirya** [comment]
> This is not true anymore after https://github.com/apache/spark/pull/42036 . It's not a big deal as we replace not-inlined CTE relations with Repartition during

`Repartition? Do you mean Relation?

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
> Repartition? Do you mean Relation?

It's Reparition, because we rely on shuffle reuse to reuse CTE relations.

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by LuciferYang** [comment]
late LGTM

**Pull request closed without merging**